src/StackExchange.Redis/AzureMaintenanceEvent.cs

```diff
@@ -0,0 +1,153 @@
+
+                    // Grab the next pair
+                    var key = message.Slice(0, message.IndexOf('|'));
+                    message = message.Slice(key.Length + 1);
+
+                    var valueEnd = message.IndexOf('|');
+                    var value = valueEnd > -1 ? message.Slice(0, valueEnd) : message;
+                    message = message.Slice(value.Length);
+
+                    if (key.Length > 0 && value.Length > 0)
+                    {
+                        switch (key)
+                        {
+                            case var _ when key.SequenceEqual(nameof(NotificationType).ToCharArray()):
+                                NotificationType = value.ToString();
+                                break;
+                            case var _ when key.SequenceEqual(nameof(StartTimeInUTC).ToCharArray()) && DateTime.TryParse(value.ToString(), out DateTime startTime):
+                                StartTimeInUTC = DateTime.SpecifyKind(startTime, DateTimeKind.Utc);
+                                break;
+                            case var _ when key.SequenceEqual(nameof(IsReplica).ToCharArray()) && bool.TryParse(value.ToString(), out var isReplica):
+                                IsReplica = isReplica;
+                                break;
+                            case var _ when key.SequenceEqual(nameof(IPAddress).ToCharArray()) && IPAddress.TryParse(value.ToString(), out var ipAddress):
+                                IpAddress = ipAddress;
+                                break;
+                            case var _ when key.SequenceEqual(nameof(SSLPort).ToCharArray()) && Int32.TryParse(value.ToString(), out var port):
+                                SSLPort = port;
+                                break;
+                            case var _ when key.SequenceEqual(nameof(NonSSLPort).ToCharArray()) && Int32.TryParse(value.ToString(), out var nonsslport):
+                                NonSSLPort = nonsslport;
+                                break;
+                            default:
+                                break;
+                        }
+                    }
+                }
+            }
+            catch
+            {
+                // TODO: Append to rolling debug log when it's present
+            }
+        }
+
+        internal async static Task AddListenerAsync(ConnectionMultiplexer multiplexer, LogProxy logProxy)
+        {
+            try
+            {
+                var sub = multiplexer.GetSubscriber();
+                if (sub == null)
+                {
+                    logProxy?.WriteLine("Failed to GetSubscriber for AzureRedisEvents");
+                    return;
+                }
+
+                await sub.SubscribeAsync(PubSubChannelName, (channel, message) =>
+                {
+                    var newMessage = new AzureMaintenanceEvent(message);
+                    multiplexer.InvokeServerMaintenanceEvent(newMessage);
```

Comment:
we can avoid the `AzureMaintenanceEvent` alloc when no subscriber, if we just pass down `message`, i.e. 

``` c#
multiplexer.InvokeServerMaintenanceEvent(message);
// ...
internal void InvokeServerMaintenanceEvent(string message)
  => AzureServerMaintenanceEvent?.Invoke(this, new AzureMaintenanceEvent(message));
```

the args on the right here are only evaluated if the `?.` check passes, i.e. we have listeners

Reply:
Even when there's no subscriber, we currently want the AzureMaintenanceEvent handy so that we can read what type of event it is and decide whether to reconfigure or not. Let me know if I understood you correctly :)

Reply:
unless I'm missing it, all we do is `AzureServerMaintenanceEvent?.Invoke` with the event; I'm just trying to avoid the overhead if `AzureServerMaintenanceEvent` turns out to be `null`, which we can reasonably presume is the default and most common scenario; do we do anything else special in the case when no external consumer has touched the event?